Branch-SDK/src/io/branch/referral/util/PromoViewHandler.java

```diff
@@ -0,0 +1,290 @@
+    private boolean isPromoDialogShowing;
+    private AppPromoView openOrInstallPendingPromo = null;
+
+    public static final int PROMO_VIEW_ERR_ALREADY_SHOWING = -200;
+
+
+    private PromoViewHandler() {
+        promoViewMap_ = new ConcurrentHashMap<>();
+    }
+
+    /**
+     * Get the singleton instance for PromoViewHandler
+     *
+     * @return {@link PromoViewHandler} instance
+     */
+    public static PromoViewHandler getInstance() {
+        if (thisInstance_ == null) {
+            thisInstance_ = new PromoViewHandler();
+        }
+        return thisInstance_;
+    }
+
+    /**
+     * Returns promo view associated with the action specified. Null if there is no Promo view available for action name
+     *
+     * @param promoAction action name
+     * @return {@link io.branch.referral.util.PromoViewHandler.AppPromoView} associated with specified action name
+     */
+    public AppPromoView getPromoView(String promoAction) {
+        boolean isPromoViewAvailable = false;
+        AppPromoView promoView = promoViewMap_.get(promoAction);
+        if (promoView != null && promoView.isAvailable()) {
+            return promoView;
+        } else {
+            return null;
+        }
+    }
+
+    public boolean showPendingPromoView(Activity currentActivity) {
+        showPromoView(openOrInstallPendingPromo, currentActivity, null);
+        return true;
+    }
+
+    public boolean showPromoView(final String action, Activity currentActivity, final IPromoViewEvents callback) {
+        if (isPromoDialogShowing) {
+            if (callback != null) {
+                callback.onPromoViewError(PROMO_VIEW_ERR_ALREADY_SHOWING, "Unable to create a promo view. A promo view is already showing");
+            }
+            return false;
+        }
+
+        isPromoDialogShowing = false;
+        AppPromoView promoView = getPromoView(action);
+
+        if (currentActivity != null && promoView != null) {
+            showPromoView(promoView, currentActivity, callback);
+            isPromoDialogShowing = true;
+        }
+
+        return isPromoDialogShowing;
+    }
+
+    private void showPromoView(final AppPromoView promoView, Activity currentActivity, final IPromoViewEvents callback) {
+        if (currentActivity != null && promoView != null) {
+            promoView.updateUsageCount();
+            isPromoDialogShowing = true;
+            WebView webView = new WebView(currentActivity);
+
+            final RelativeLayout layout = new RelativeLayout(currentActivity);
+            layout.setVisibility(View.GONE);
+
+            RelativeLayout.LayoutParams layoutParams = new RelativeLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT);
+            layoutParams.setMargins(40, 80, 40, 80);
+
+            layout.addView(webView, layoutParams);
+            layout.setBackgroundColor(Color.parseColor("#11FEFEFE"));
+
+            final TextView confirmTxt = new TextView(currentActivity);
```

Comment:
Please replace these buttons with in-HTML confirm + cancel URL interception. I'd vote: `branch-cta://cancel` + `branch-cta://confirm`